### PR TITLE
(closes #1173) reshape LFRic mesh map

### LINKED ISSUE #1173: Reshaping of LFRic intergrid mesh maps
As part of LFRic ticket 2572 we are reshaping the mesh map that encodes which cells on a fine mesh are associated with which cells on a coarse mesh.

The idea is that the mesh map won't just be a list of fine cells per coarse cell, instead it will be a table whose layout matches the arrangement of the fine cells within the coarse cell.

To facilitate this any LFRic intergrid kernels would now take `ncell_f_per_c_x` and `ncell_f_per_c_y` as arguments instead of just `ncell_f_per_c`, e.g.:

```
  subroutine restrict_kernel_code(nlayers,         &
                                  cell_map,        &
                                  ncell_f_per_c_x, &
                                  ncell_f_per_c_y, &
                                  ncell_f,         &
                                  coarse,          &
                                  fine,            &
                                  undf_c,          &
                                  dofmap_c,        &
                                  ndf,             &
                                  undf_f,          &
                                  dofmap_f)
```

I think we'll need to update the following bit of Psyclone for this to be supported:
https://github.com/stfc/PSyclone/blob/9d6919083057c48c437010ba73e18aad8a32eeb6/src/psyclone/domain/lfric/kern_call_arg_list.py#L106-L109

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is now ready for first review. Probably from @TeranIvy or @rupertford :-)

**Comment by TeranIvy** [comment]
Happy to take it, @arporter.

File: src/psyclone/dynamo0p3.py [hunk 1]
```diff
@@ -4045,9 +4054,12 @@ def __init__(self, fine_arg, coarse_arg):
...
             "ncell_{0}".format(fine_arg.name)).name
-        # No. of fine cells per coarse cell
-        self.ncellpercell = symtab.symbol_from_tag(
-            "ncpc_{0}_{1}".format(fine_arg.name, coarse_arg.name)).name
+        # No. of fine cells per coarse cell in x
+        self.ncellpercellx = symtab.symbol_from_tag(
+            "ncpc_{0}_{1}_x".format(fine_arg.name, coarse_arg.name)).name
+        # No. of fine cells per coarse cell in y
+        self.ncellpercelly = symtab.symbol_from_tag(
+            "ncpc_{0}_{1}_y".format(fine_arg.name, coarse_arg.name)).name
```
Comment (TeranIvy) [inline src/psyclone/dynamo0p3.py:4062]: This is not the fault of this PR, but `pylint` reports
```
dynamo0p3.py:7167:0: C0330: Wrong hanging indentation (remove 7 spaces).
                                          prev_node_name)))
                                   |      ^ (bad-continuation)
```
Would you mind fixing this?

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/mesh_map_mod.F90 [hunk 2]
```diff
@@ -61,7 +61,7 @@ module mesh_map_mod
 
   private
 
-  integer(i_def), allocatable :: mesh_map(:,:) ! In LIDs
+  integer(i_def), allocatable :: mesh_map(:,:,:) ! In LIDs
```
Comment (TeranIvy) [inline src/psyclone/tests/test_files/dynamo0p3/infrastructure/mesh_map_mod.F90:64]: Please change "2020" to "2021" in the file Copyrights.

File: doc/user_guide/dynamo0p3.rst [hunk 3]
```diff
@@ -1957,12 +1957,14 @@ arguments to inter-grid kernels are as follows:
    is an integer of type ``i_def`` and has intent ``in``.
 
 2) Include the ``cell_map`` for the current cell (column). This is
-   an integer array of rank one, type ``i_def`` and intent ``in``
+   an integer array of rank two, type ``i_def`` and intent ``in``
    which provides the mapping from the coarse to the fine mesh. It
-   has extent `ncell_f_per_c`.
+   has extent ``ncell_f_per_c_x, ncell_f_per_c_y``.
```
Comment (TeranIvy) [inline doc/user_guide/dynamo0p3.rst:1962]: I would suggest enclosing this tuple in brackets, i.e. ``(ncell_f_per_c_x, ncell_f_per_c_y)``.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
This is a nice, self-contained piece of work.

The changes are `pycodestyle` and `pylint` clean. There is one `pylint` complaint for `dynamo0p3.py` not related to this PR, but since I am requesting changes I would kindly ask the developer to fix it here (see inline). There is also a copyright year that needs to be updated (see inline).

`pytest` tests run with compilation. All changes are covered.

Documentation builds and looks fine, both PDF and HTML. I have a minor suggestion for the LFRic API documentation (see inline).

Examples run fine, however the integrid kernels in `examples/lfric/eg10`  **will have to be updated as well**. I would suggest using the LFRic trunk diffs (linked), for the [prolong](https://code.metoffice.gov.uk/trac/lfric/changeset?reponame=&new=28090%40LFRic%2Ftrunk%2Fgungho%2Fsource%2Fkernel%2Fprolong_kernel_mod.F90&old=27470%40LFRic%2Ftrunk%2Fgungho%2Fsource%2Fkernel%2Fprolong_kernel_mod.F90) and [restrict](https://code.metoffice.gov.uk/trac/lfric/changeset?reponame=&new=28090%40LFRic%2Ftrunk%2Fgungho%2Fsource%2Fkernel%2Frestrict_kernel_mod.F90&old=27470%40LFRic%2Ftrunk%2Fgungho%2Fsource%2Fkernel%2Frestrict_kernel_mod.F90) kernels, respectively.

The changes were tested against LFRic trunk on both Met Office Spice cluster and Cray XC40 machines with a PSyclone environment built from this branch. I have removed the PSyKAl-lite calls to the ``prolong`` and ``restrict`` kernels in multiple ``gungho`` algorithms and ran the full test suite on the ``gungho``, ``lfric_atm`` and ``gravity_wave`` applications. I am happy to report that everything is fine.

I would also kindly ask the developer to bring the branch up-to-date with the master.

File: src/psyclone/dynamo0p3.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/dynamo0p3.py:4062]: I couldn't get my pylint to complain about that (2.7.4) but I think (hope) I've fixed it.

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/mesh_map_mod.F90 (hunk 2 above)
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/mesh_map_mod.F90:64]: Done.

File: doc/user_guide/dynamo0p3.rst (hunk 3 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:1962]: Yes, that's better.

**Comment by arporter** [comment]
Ready for another look now.

**Comment by arporter** [comment]
Taking this back for a mo to update kernel metadata.

**Comment by arporter** [comment]
Assuming the CI is happy, this is ready now.

File: src/psyclone/dynamo0p3.py (hunk 1 above)
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:4062]: It's fine now, thanks.

**Review by TeranIvy — APPROVED** [review]
All well now. Tests with compilation pass, examples run and documentation builds. The changes are `pycodestyle` and `pylint` clean.
Updating the `lfric/eg10` kernels reverted them to the old-style metadata but I'm happy to fix this in #874.

Updated the changelog and User Guide. Proceeding to merge assuming that Travis is happy.

**Comment by arporter** [comment]
I took the liberty of merging this so that I can update my branch in #870, hope that's OK.

**Comment by TeranIvy** [comment]
Ok, we'll race between #874 and #870 then.

**Pull request merged**